### ISSUE TITLE
timeouts with fetch_adjacencies

### ISSUE THREAD

**Issue by schlegelp** [connectome-neuprint/neuprint-python#7]
The recent change in `fetch_adjacencies` to batch only the smaller set of neurons between `sources` and `targets` broke larger queries.

E.g. I am trying to fetch adjacencies between ~300 source and ~3000 target neurons. This leads to a huge list of bodyIds in `WHERE m.bodyId in [12356, ...]` which almost always results in timeouts. Fetching __all__ downstream targets of those 300 source neurons in batches without any target restrictions works like a charm though. 

I appreciate that batching either sources or targets but not both makes thing faster but sometimes slow but steady wins the race. I see a few options:
- go back to batching both sources and targets
- turn things around and always batch the larger set of neurons between `sources` and `targets` (might not solve the issue if both lists are large)
- implement some smart logic for batching
- give a warning if larger lists is above a certain length and suggest to just grab everything

Thoughts?

**Comment by stuarteberg** [comment]
Thanks for pointing out this problem.  I was able to reproduce the timeout you describe, and released `0.4.6` to (hopefully) fix the issue.

**Details:**

The most concerning thing is that your queries are taking more than a few seconds, let alone timing out.  They really shouldn't take so long.

I spent some time experimenting with the relevant cypher query, and also debugging in the neo4j web UI.  I ended up with the following changes:

- Large body lists (or type/instance lists) will be pre-defined at the top of the query, in a `WITH` clause.
- That improves things somewhat, but strangely, neo4j still does not produce an optimal query.  The query planner introduces a `CartesianProduct` operation where none is necessary.  However, the planner can be tricked into doing the right thing -- by introducing a seemingly pointless `WITH` statement (with a dummy variable) between the `WHERE` clause for the source criteria and the `WHERE` clause for the target criteria.  That improves the performance significantly.

The end result looks approximately like this:

   ```
    WITH [202916528, 203253072, ...] as n_search_bodyIds,
         [549385850, 549744594, ...] as m_search_bodyIds

    MATCH (n:Neuron)-[e:ConnectsTo]->(m:Neuron)
    WHERE m.bodyId in m_search_bodyIds

    // Artificial break in the query flow to fool the query planner
    WITH *, true as _
    WHERE n.bodyId in n_search_bodyIds

    RETURN n.bodyId as bodyId_pre,
           m.bodyId as bodyId_post,
           e.weight as weight,
           e.roiInfo as roiInfo
```

These changes complicate the `neuprint-python` code in some places, but the performance benefits are significant, so I went ahead and committed them and released `0.4.6`.  Let me know if the new version works for you.

**Comment by schlegelp** [comment]
Works like a charm now, thanks!

**Comment by stuarteberg** [comment]
Thanks for trying out the fix.

**Comment by stuarteberg** [comment]
Note to myself, for later:

Steve pointed out that there's also a different way to fool the query planner.  The trick is not to let the query planner use one of the indexes, so it can't attempt a `CartesianProduct` at all.  That is achieved by leaving out `:Neuron` from one of the match variable definitions.  The presence of a `:Neuron` label can be enforced later in the query, in the `WHERE` clause.

```
 WITH [202916528, 203253072, ...] as n_search_bodyIds,
      [549385850, 549744594, ...] as m_search_bodyIds

 MATCH (n)-[e:ConnectsTo]->(m:Neuron)
 WHERE n.Neuron
      AND m.bodyId in m_search_bodyIds
      AND n.bodyId in n_search_bodyIds

 RETURN n.bodyId as bodyId_pre,
        m.bodyId as bodyId_post,
        e.weight as weight,
        e.roiInfo as roiInfo
```

**Comment by jefferis** [comment]
Just a query on this @stuarteberg. In http://github.com/natverse/neuprintr the query looks like this:

```
WITH %s AS input, %s AS output MATCH (n:Neuron)-[c:ConnectsTo]->(m)
      WHERE n.bodyId IN input AND m.bodyId IN output
      RETURN n.bodyId AS upstream, m.bodyId AS downstream, c.weight AS weight, 
      n.instance AS upName, m.instance AS downName
```
(obvious %s replacements)

So 
1. it seems like we never had this issue because we were always using a connection with one end not specified as a Neuron just as in Steve's sample query above.
2. I'm not sure why it's necessary to add the `WHERE n.Neuron` since I would assume that it's up to the user to decide what bodyids to supply as input to the query. But if one is going to restrict, shouldn't one do so for m and n?

**Comment by stuarteberg** [comment]
>I'm not sure why it's necessary to add the WHERE n.Neuron since I would assume that it's up to the user to decide what bodyids to supply as input to the query. 

You're right, in most cases it's highly unlikely to make a difference in most use-cases.  The reason I added it is somewhat subtle, and specific to the design of new `neuprint-python` API, perhaps not relevant to `neuprintr`.  But read on if you are interested.

In `neuprint-python`, users can specify their neurons of interest by providing [`NeuronCriteria`][1].  The semantics are such that every field in a `NeuronCriteria` applies an additional filter, *narrowing* the returned set of bodies.  Basically, all the conditions you specify in the criteria are joined together with `AND`.**

For example:

```python
from neuprint import NeuronCriteria as NC
nc = NC(type=["EL","EPG"], status="Traced", label="Neuron")
synapses = fetch_synapses(nc)
```

For consistency with these rules, the same logic is applied to the `bodyId` field -- it can only narrow the search, not expand it.

```python
from neuprint import NeuronCriteria as NC, fetch_synapses
nc = NC(bodyId=[1292882359, 1036753721,  820144012],
        type=["EL","EPG"], status="Traced", label="Neuron")
synapses = fetch_synapses(nc)
```

But I agree, it's unlikely to make a difference, since most users probably won't use the `bodyId` field AND the other search fields at the same time.  I imagine they'll use one or the other.

>But if one is going to restrict, shouldn't one do so for m and n?

I'm not a cypher expert, but my reasoning was that I want to enforce the `Neuron` requirement on both `n` and `m`, and in cypher, labels can be checked in either the `MATCH` clause or the `WHERE` clause.  My initial query placed the requirement in the `MATCH` clause for both `n` and `m`, but for performance reasons, I moved one of the requirements into the `WHERE` clause instead.  But there's no need to add the other to the `WHERE` clause -- it remains in the `MATCH`, so adding it to the `WHERE` clause would be redundant.

[1]: https://connectome-neuprint.github.io/neuprint-python/docs/neuroncriteria.html
